// object-settings/static/lib-cov/object-settings.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['object-settings.js']) {
  _$jscoverage['object-settings.js'] = [];
  _$jscoverage['object-settings.js'][8] = 0;
  _$jscoverage['object-settings.js'][11] = 0;
  _$jscoverage['object-settings.js'][12] = 0;
  _$jscoverage['object-settings.js'][15] = 0;
  _$jscoverage['object-settings.js'][16] = 0;
  _$jscoverage['object-settings.js'][17] = 0;
  _$jscoverage['object-settings.js'][20] = 0;
  _$jscoverage['object-settings.js'][32] = 0;
  _$jscoverage['object-settings.js'][33] = 0;
  _$jscoverage['object-settings.js'][35] = 0;
  _$jscoverage['object-settings.js'][37] = 0;
  _$jscoverage['object-settings.js'][46] = 0;
  _$jscoverage['object-settings.js'][62] = 0;
  _$jscoverage['object-settings.js'][64] = 0;
  _$jscoverage['object-settings.js'][67] = 0;
  _$jscoverage['object-settings.js'][68] = 0;
  _$jscoverage['object-settings.js'][71] = 0;
  _$jscoverage['object-settings.js'][73] = 0;
  _$jscoverage['object-settings.js'][85] = 0;
  _$jscoverage['object-settings.js'][86] = 0;
  _$jscoverage['object-settings.js'][89] = 0;
  _$jscoverage['object-settings.js'][99] = 0;
  _$jscoverage['object-settings.js'][110] = 0;
  _$jscoverage['object-settings.js'][111] = 0;
  _$jscoverage['object-settings.js'][120] = 0;
  _$jscoverage['object-settings.js'][121] = 0;
  _$jscoverage['object-settings.js'][124] = 0;
  _$jscoverage['object-settings.js'][127] = 0;
  _$jscoverage['object-settings.js'][142] = 0;
  _$jscoverage['object-settings.js'][143] = 0;
  _$jscoverage['object-settings.js'][152] = 0;
  _$jscoverage['object-settings.js'][153] = 0;
  _$jscoverage['object-settings.js'][157] = 0;
  _$jscoverage['object-settings.js'][160] = 0;
  _$jscoverage['object-settings.js'][163] = 0;
  _$jscoverage['object-settings.js'][165] = 0;
  _$jscoverage['object-settings.js'][170] = 0;
  _$jscoverage['object-settings.js'][172] = 0;
  _$jscoverage['object-settings.js'][175] = 0;
  _$jscoverage['object-settings.js'][176] = 0;
  _$jscoverage['object-settings.js'][177] = 0;
  _$jscoverage['object-settings.js'][183] = 0;
  _$jscoverage['object-settings.js'][194] = 0;
  _$jscoverage['object-settings.js'][195] = 0;
  _$jscoverage['object-settings.js'][203] = 0;
  _$jscoverage['object-settings.js'][214] = 0;
  _$jscoverage['object-settings.js'][220] = 0;
  _$jscoverage['object-settings.js'][221] = 0;
}
_$jscoverage['object-settings.js'][8]++;
require("abstract").createModel(module.exports = (function Settings(exports) {
  _$jscoverage['object-settings.js'][11]++;
  if (module.loaded) {
    _$jscoverage['object-settings.js'][12]++;
    return Settings;
  }
  _$jscoverage['object-settings.js'][15]++;
  var util = require("util");
  _$jscoverage['object-settings.js'][16]++;
  var inherits = require("util").inherits;
  _$jscoverage['object-settings.js'][17]++;
  var _extend = require("util")._extend;
  _$jscoverage['object-settings.js'][20]++;
  Settings.defineProperties(Settings, {inject: {value: (function inject(obj) {
  _$jscoverage['object-settings.js'][32]++;
  var target = obj || this;
  _$jscoverage['object-settings.js'][33]++;
  var Instance = new Settings.create();
  _$jscoverage['object-settings.js'][35]++;
  Object.getOwnPropertyNames(Settings.prototype).forEach((function (key) {
  _$jscoverage['object-settings.js'][37]++;
  Object.defineProperty(target, key, {value: Instance[key], enumerable: false, writable: true, configurable: true});
}));
  _$jscoverage['object-settings.js'][46]++;
  return target;
}), enumerable: true, configurable: true, writable: true}, mixin: {value: (function mixin(obj) {
  _$jscoverage['object-settings.js'][62]++;
  for (var key in Settings.prototype) {
    _$jscoverage['object-settings.js'][64]++;
    var descriptor = Object.getOwnPropertyDescriptor(obj, key);
    _$jscoverage['object-settings.js'][67]++;
    if (descriptor && ! descriptor.configurable) {
      _$jscoverage['object-settings.js'][68]++;
      break;
    }
    _$jscoverage['object-settings.js'][71]++;
    try {
      _$jscoverage['object-settings.js'][73]++;
      Object.defineProperty(obj, key, {value: obj[key] || Settings.prototype[key], enumerable: false, writable: true, configurable: true});
    }
    catch (error) {
    }
}
  _$jscoverage['object-settings.js'][85]++;
  if (! obj._events) {
    _$jscoverage['object-settings.js'][86]++;
    obj._events = {};
  }
  _$jscoverage['object-settings.js'][89]++;
  return obj;
}), enumerable: true, configurable: true, writable: true}});
  _$jscoverage['object-settings.js'][99]++;
  Settings.properties(Settings.prototype, {get: (function get(key, fallback) {
  _$jscoverage['object-settings.js'][110]++;
  if (! this._meta) {
    _$jscoverage['object-settings.js'][111]++;
    Object.defineProperty(this, "_meta", {value: {}, enumerable: false, writable: true, configurable: true});
  }
  _$jscoverage['object-settings.js'][120]++;
  if ("undefined" === typeof key) {
    _$jscoverage['object-settings.js'][121]++;
    return this._meta || {};
  }
  _$jscoverage['object-settings.js'][124]++;
  var value = require("abstract").utility.query(this._meta, arguments[0]);
  _$jscoverage['object-settings.js'][127]++;
  return value || fallback;
}), set: (function set(key, value) {
  _$jscoverage['object-settings.js'][142]++;
  if (! this._meta) {
    _$jscoverage['object-settings.js'][143]++;
    Object.defineProperty(this, "_meta", {value: {}, enumerable: false, writable: true, configurable: true});
  }
  _$jscoverage['object-settings.js'][152]++;
  if (! arguments) {
    _$jscoverage['object-settings.js'][153]++;
    return this;
  }
  _$jscoverage['object-settings.js'][157]++;
  var emit = this.emit? this.emit.bind(this): (function emit() {
});
  _$jscoverage['object-settings.js'][160]++;
  if (Object.keys(arguments).length === 2 && ("string" === typeof arguments[0] || "number" === typeof arguments[0])) {
    _$jscoverage['object-settings.js'][163]++;
    require("abstract").utility.unwrap(arguments[0], arguments[1], this._meta);
    _$jscoverage['object-settings.js'][165]++;
    emit(["set", arguments[0]], null, arguments[1], arguments[0]);
  }
  _$jscoverage['object-settings.js'][170]++;
  if (Object.keys(arguments).length === 1 && "object" === typeof arguments[0]) {
    _$jscoverage['object-settings.js'][172]++;
    require("util")._extend(this._meta, arguments[0]);
    _$jscoverage['object-settings.js'][175]++;
    for (var key in arguments[0]) {
      _$jscoverage['object-settings.js'][176]++;
      if (arguments[0].hasOwnProperty(key)) {
        _$jscoverage['object-settings.js'][177]++;
        emit(["set", key], null, arguments[0][key], key);
      }
}
  }
  _$jscoverage['object-settings.js'][183]++;
  return this;
}), enable: (function enable(key) {
  _$jscoverage['object-settings.js'][194]++;
  if (! this._meta) {
    _$jscoverage['object-settings.js'][195]++;
    Object.defineProperty(this, "_meta", {value: {}, enumerable: false, writable: true, configurable: true});
  }
  _$jscoverage['object-settings.js'][203]++;
  this._meta[key] = true;
}), disable: (function disable(key) {
  _$jscoverage['object-settings.js'][214]++;
  return this._meta? this._meta[key] = false: null;
})});
  _$jscoverage['object-settings.js'][220]++;
  Settings.defineConstructor((function (defaults) {
  _$jscoverage['object-settings.js'][221]++;
  this.set(defaults);
}));
}));
_$jscoverage['object-settings.js'].source = ["/**"," * Create custom objects and methods by aggregating and abstracting esources."," *"," * @version 0.1.0"," * @class Settings"," * @constractor"," */","require( 'abstract' ).createModel( module.exports = function Settings( exports ) {","","  // Construct Model only once.","  if( module.loaded ) {","    return Settings;","  }","","  var util       = require( 'util' );","  var inherits   = require( 'util' ).inherits;","  var _extend    = require( 'util' )._extend;","","  // Constructor Properties","  Settings.defineProperties( Settings, {","    inject: {","      /**","       * Force / Override properties","       *","       * If object not provided will bind to context.","       *","       * @param {Object} obj","       * @return {Object}","       */","      value: function inject( obj ) {","","        var target = obj || this;","        var Instance = new Settings.create();","","        Object.getOwnPropertyNames( Settings.prototype ).forEach( function( key ) {","","          Object.defineProperty( target, key, {","            value: Instance[ key ],","            enumerable: false,","            writable: true,","            configurable: true","          });","","        });","","        return target;","","      },","      enumerable: true,","      configurable: true,","      writable: true","    },","    mixin: {","      /**","       * Mixin the Emitter properties.","       *","       * @param {Object} obj","       * @return {Object}","       */","      value: function mixin( obj ) {","","        for( var key in Settings.prototype ) {","","          var descriptor = Object.getOwnPropertyDescriptor( obj, key );","","          // Detect if a property is not configurable.","          if( descriptor &amp;&amp; !descriptor.configurable ) {","            break;","          }","","          try {","","            Object.defineProperty( obj, key, {","              value: obj[ key ] || Settings.prototype[ key ],","              enumerable: false,","              writable: true,","              configurable: true","            });","","          } catch( error ) {}","","        }","","        // Ensure we have an listener container.","        if( !obj._events ) {","          obj._events = {};","        }","","        return obj;","","      },","      enumerable: true,","      configurable: true,","      writable: true","    }","  });","","  // Prototypal Properties","  Settings.properties( Settings.prototype, {","","    /**","     * Get or create and get storage","     *","     * @param key {String}","     * @returns {*|undefined}","     */","    get: function get( key, fallback ) {","","      // Create object meta if it does not exist","      if( !this._meta ) {","        Object.defineProperty( this, '_meta', {","          value: {},","          enumerable: false,","          writable: true,","          configurable: true","        });","      }","","      // Return empty full meta object if no key specified","      if( 'undefined' === typeof key ) {","        return this._meta || {};","      }","","      var value = require( 'abstract' ).utility.query( this._meta, arguments[0] );","","      // Return the found value or the fallback value","      return value || fallback;","","    },","","    /**","     * Set Key &amp; Value pair, or pass an object","     *","     * @method key","     * @param key {String|Object}","     * @param value {Any}","     * @returns {Object} Context.","     */","    set: function set( key, value ) {","","      // Create object meta if it does not exist","      if( !this._meta ) {","        Object.defineProperty( this, '_meta', {","          value: {},","          enumerable: false,","          writable: true,","          configurable: true","        });","      }","","      // Not passing any arguments can be used to instantiate","      if( !arguments ) {","        return this;","      }","","      // Wrapper for Emit","      var emit = this.emit ? this.emit.bind( this ) : function emit() {};","","      // Key &amp; Value Passed","      if( Object.keys( arguments ).length === 2 &amp;&amp; ( 'string' === typeof arguments[0] || 'number' === typeof arguments[0] ) ) {","","        // Honor dot notation","        require( 'abstract' ).utility.unwrap( arguments[0], arguments[1], this._meta );","","        emit( [ 'set', arguments[0] ], null, arguments[1], arguments[0] );","","      }","","      // Object Passed, extend","      if( Object.keys( arguments ).length === 1 &amp;&amp; 'object' === typeof arguments[0] ) {","","        require( 'util' )._extend( this._meta, arguments[0] );","","        // @todo Get object path using dot notation","        for( var key in arguments[0] ) {","          if( arguments[0].hasOwnProperty( key ) ) {","            emit( [ 'set', key ], null, arguments[0][key], key );","          }","        }","","      }","","      return this;","","    },","","    /**","     * Enable an Option","     *","     * @param key","     */","    enable: function enable( key ) {","","      if( !this._meta ) {","        Object.defineProperty( this, '_meta', {","          value: {},","          enumerable: false,","          writable: true,","          configurable: true","        });","      }","","      this._meta[ key ] = true;","","    },","","    /**","     * Disable an Option","     *","     * @param key","     * @returns {boolean}","     */","    disable: function disable( key ) {","      return this._meta ? this._meta[ key ] = false : null;","    }","","  });","","  // Define instance constructor and bind to module.exports","  Settings.defineConstructor( function( defaults ) {","    this.set( defaults );","  });","","});",""];
